pkg/client/client_venconn_test.go

```diff
@@ -0,0 +1,381 @@
+// controller-runtime's fake clientset doesn't support server-side apply [1] and
+// also because we want to create serviceaccount tokens, which isn't supported
+// by the fake clientset either.
+//
+// The goal is to test the following behaviors:
+//
+//   - VenafiConnection's `accessToken` works as expected with a fake Venafi
+//     Cloud server.
+//   - VenafiConnection's `apiKey` and `tpp` can't be used by the user.
+//   - NewVenConnClient's `trustedCAs` works as expected.
+//
+// [1] https://github.com/kubernetes-sigs/controller-runtime/issues/2341
+func TestVenConnClient_PostDataReadingsWithOptions(t *testing.T) {
+	t.Parallel()
+
+	t.Run("valid accessToken", run(testcase{
+		given: undent(`
+			apiVersion: jetstack.io/v1alpha1
+			kind: VenafiConnection
+			metadata:
+			  name: venafi-components
+			  namespace: venafi
+			spec:
+			  vcp:
+			    url: FAKE_VENAFI_CLOUD_URL
+			    accessToken:
+			      - secret:
+			          name: accesstoken
+			          fields: [accesstoken]`),
+		expectReadyCondMsg: "ea744d098c2c1c6044e4c4e9d3bf7c2a68ef30553db00f1714886cedf73230f1",
+	}))
+	t.Run("error when the apiKey field is used", run(testcase{
+		// Why isn't it possible to use the 'apiKey' field? Although the
+		// Kubernetes Discovery endpoint works with an API key, we have decided
+		// to not support it because it isn't recommended.
+		given: undent(`
+			apiVersion: jetstack.io/v1alpha1
+			kind: VenafiConnection
+			metadata:
+			  name: venafi-components
+			  namespace: venafi
+			spec:
+			  vcp:
+			    url: FAKE_VENAFI_CLOUD_URL
+			    apiKey:
+			      - secret:
+			          name: apikey
+			          fields: [apikey]`),
+		expectReadyCondMsg: "b099d634ccec56556da28028743475dab67f79d079b668bedc3ef544f7eed2f3",
+		expectErr:          "VenafiConnection venafi/venafi-components: the agent cannot be used with an API key",
+	}))
+	t.Run("error when the tpp field is used", run(testcase{
+		// IMPORTANT: The user may think they can use 'tpp', spend time
+		// debugging and making the venafi connection work, and then find out
+		// that it doesn't work. The reason is because as of now, we don't first
+		// check if the user has used the 'tpp' field before running Get.
+		given: undent(`
+			apiVersion: jetstack.io/v1alpha1
+			kind: VenafiConnection
+			metadata:
+			  name: venafi-components
+			  namespace: venafi
+			spec:
+			  tpp:
+			    url: FAKE_TPP_URL
+			    accessToken:
+			      - secret:
+			          name: accesstoken
+			          fields: [accesstoken]`),
+		expectErr:          ``,
+		expectReadyCondMsg: `ea744d098c2c1c6044e4c4e9d3bf7c2a68ef30553db00f1714886cedf73230f1`,
+	}))
+}
+
+type testcase struct {
+	given              string
+	expectErr          string
+	expectReadyCondMsg string
+}
+
+func run(test testcase) func(t *testing.T) {
+	return func(t *testing.T) {
+		fakeVenafiCloud, certCloud := fakeVenafiCloud(t)
+		fakeTPP, certTPP := fakeTPP(t)
+		_, restconf, kclient := startEnvtest(t)
```

Comment:
The returned envtest is discarded here, but I think you should `t.Cleanup(env.Stop)`. I see envtest processes lingering after I run these tests

```console
$ pidof etcd kube-apiserver
374054 373960
```

Reply:
That's odd. startEnvtest uses `t.Cleanup` to stop the kube-apiserver and etcd:

```go
func startEnvtest(t testing.TB) (_ *envtest.Environment, _ *rest.Config, kclient ctrlruntime.WithWatch) {
	// ...
	t.Cleanup(func() {
		t.Log("Waiting for envtest to exit")
		err = envtest.Stop()
		require.NoError(t, err)
	})
```

Not sure why it isn't cleaned up. I'll have to investigate. Would it be OK if we merge this PR even if this oddity is still there?